File: examples/cpptest/src/main/cpp/subsystems/DiffDriveSubsystem.cpp
// Copyright (c) 2026 Yet Another Software Suite
// SPDX-License-Identifier: LGPL-3.0-or-later

#include "subsystems/DiffDriveSubsystem.h"

#include <frc/system/plant/DCMotor.h>
#include <units/length.h>

using namespace yams::motorcontrollers;
using namespace yams::gearing;
using Cfg = SmartMotorControllerConfig;

DiffDriveSubsystem::DiffDriveSubsystem() {
  // 3:1 then 4:1 in series = 12:1 total rotor-to-wheel reduction.
  // 4-inch wheel diameter (4.0 * 0.0254 m) gives circumference used for distance tracking.
  MechanismGearing gearing{GearBox::FromReductionStages({3.0, 4.0})};
  const units::meter_t wheelDiameter{4.0 * 0.0254};

  // Left side inverted because the left motor is mounted mirrored to the right.
  // Open-loop mode: DifferentialDrive sends duty-cycle directly, no closed-loop PID needed
  // for basic tank/arcade drive.  COAST idle mode: wheels roll freely when the driver
  // releases the stick, which feels natural for driver practice but may overshoot in auto --
  // switch to BRAKE if you want the robot to stop sharply.
  //
  // DCMotor::NEO(2) tells the sim there are 2 NEOs per side (two motors per gearbox).
  // This affects the simulated stall torque and free speed.
  m_leftConfig.WithSubsystem(this)
      .WithOpenLoopMode()
      .WithMotorGearing(gearing)
      .WithIdleMode(Cfg::MotorMode::COAST)
      .WithMotorInverted(true)
      .WithMechanismCircumference(wheelDiameter * 3.14159265)
      .WithTelemetry("LeftMotorMain", Cfg::TelemetryVerbosity::LOW);

  m_rightConfig.WithSubsystem(this)
      .WithOpenLoopMode()
      .WithMotorGearing(gearing)
      .WithIdleMode(Cfg::MotorMode::COAST)
      .WithMotorInverted(false)
      .WithMechanismCircumference(wheelDiameter * 3.14159265)
      .WithTelemetry("RightMotorMain", Cfg::TelemetryVerbosity::LOW);

  // SparkWrapper takes a pointer (CAN IDs 21 and 24 from the header).
  // Followers (22 and 23) are not wired through YAMS here -- configure them via
  // SmartMotorControllerConfig::WithFollowers if you need YAMS to manage them.
  m_leftSMC.emplace(&m_leftMotor, frc::DCMotor::NEO(2), &m_leftConfig);
  m_rightSMC.emplace(&m_rightMotor, frc::DCMotor::NEO(2), &m_rightConfig);

  // DifferentialDrive's output callbacks: each lambda captures the SparkWrapper by pointer
  // and forwards the duty-cycle from WPILib's tank/arcade math to the actual motor.
  // The callbacks are invoked by DifferentialDrive::TankDrive / ArcadeDrive every loop --
  // they must stay valid for the lifetime of m_drive, which is why they capture `this`.
  m_drive.emplace([this](double dc) { m_leftSMC->SetDutyCycle(dc); },
                  [this](double dc) { m_rightSMC->SetDutyCycle(dc); });

  // SetDefaultCommand(Stop()) ensures the robot applies zero output whenever no other
  // command is running on this subsystem.  Without a default command the motors would
  // hold their last duty-cycle across command boundaries, which can cause unintended
  // movement or brownouts during auto-to-teleop transitions.
  SetDefaultCommand(Stop());
}

// Actively commands 0% output each loop.  Using Run (not RunOnce) means the subsystem
// is continuously required, so any command that needs the drivetrain will properly
// interrupt Stop rather than fighting it.
frc2::CommandPtr DiffDriveSubsystem::Stop() {
  return Run([this] { m_drive->StopMotor(); });
}

// Supplier-based overloads let callers pass joystick lambdas; the values are read
// fresh each loop inside the Run closure so axis changes are picked up immediately.
frc2::CommandPtr DiffDriveSubsystem::TankDrive(std::function<double()> left,
                                               std::function<double()> right) {
  return Run([this, left, right] { m_drive->TankDrive(left(), right()); });
}

// zRotation is the turning rate (positive = turn left per WPILib convention).
frc2::CommandPtr DiffDriveSubsystem::ArcadeDrive(std::function<double()> xSpeed,
                                                 std::function<double()> zRotation) {
  return Run([this, xSpeed, zRotation] { m_drive->ArcadeDrive(xSpeed(), zRotation()); });
}

void DiffDriveSubsystem::Periodic() {}

// Steps each side's DCMotorSim independently.  Sim does not model follower motors;
// each SparkWrapper simulates only the single motor passed to its constructor.
void DiffDriveSubsystem::SimulationPeriodic() {
  m_leftSMC->SimIterate();
  m_rightSMC->SimIterate();
}

File: examples/Kitbot2026/java/frc/robot/subsystems/DiffDriveSubsystem.java
// Copyright (c) 2026 Yet Another Software Suite
// SPDX-License-Identifier: LGPL-3.0-or-later

package frc.robot.subsystems;


import static edu.wpi.first.units.Units.Inches;

import com.revrobotics.spark.SparkMax;
import edu.wpi.first.math.Pair;
import edu.wpi.first.math.system.plant.DCMotor;
import edu.wpi.first.units.measure.Distance;
import edu.wpi.first.wpilibj.drive.DifferentialDrive;
import edu.wpi.first.wpilibj2.command.Command;
import edu.wpi.first.wpilibj2.command.SubsystemBase;
import java.util.function.DoubleSupplier;
import yams.gearing.MechanismGearing;
import yams.motorcontrollers.SmartMotorController;
import yams.motorcontrollers.SmartMotorControllerConfig;
import yams.motorcontrollers.SmartMotorControllerConfig.ControlMode;
import yams.motorcontrollers.SmartMotorControllerConfig.MotorMode;
import yams.motorcontrollers.SmartMotorControllerConfig.TelemetryVerbosity;
import yams.motorcontrollers.local.SparkWrapper;

/**
 * Tank-drive base for the 2026 Kitbot. Four NEOs in a 2+2 leader/follower layout
 * (two per side). OPEN_LOOP only -- no encoder-based velocity hold -- which is
 * fine for driver control where the human provides the outer feedback loop.
 *
 * <p>Motor IDs follow the Kitbot wiring convention:
 *   Left  leader=21, follower=22
 *   Right leader=24, follower=23
 *
 * <p>arcadeDrive is set as the default command so the drive always has a
 * safe fallback if no other command claims it.
 */
public class DiffDriveSubsystem extends SubsystemBase
{
  // 3:4 reduction -- Kitbot gearbox; two NEOs per side so DCMotor.getNEO(2) models
  // both motors on one side as a single gearbox input for simulation accuracy.
  private MechanismGearing gearing = new MechanismGearing(3, 4);

  // 4-inch wheels are the standard Kitbot wheel size.
  private Distance         wheelDiameter = Inches.of(4);

  private SparkMax leftMotor  = new SparkMax(21, SparkMax.MotorType.kBrushless);
  private SparkMax rightMotor = new SparkMax(24, SparkMax.MotorType.kBrushless);

  // Followers mirror the leader output. false = same direction as leader.
  private SparkMax leftFollowerMotor  = new SparkMax(22, SparkMax.MotorType.kBrushless);
  private SparkMax rightFollowerMotor = new SparkMax(23, SparkMax.MotorType.kBrushless);

  // Left side is inverted because the motors are physically mirrored on the
  // chassis -- positive voltage on both sides would otherwise spin them in
  // opposite directions.
  private SmartMotorControllerConfig leftMotorConfig = new SmartMotorControllerConfig(this)
      .withControlMode(ControlMode.OPEN_LOOP)
      .withGearing(gearing)
      .withIdleMode(MotorMode.COAST) // COAST lets the robot roll to a stop; easier to push around when disabled
      .withMotorInverted(true)
      .withWheelDiameter(wheelDiameter)
      .withTelemetry("LeftMotorMain", TelemetryVerbosity.LOW)
      .withFollowers(Pair.of(leftFollowerMotor, false)); // follower not inverted relative to leader

  private SmartMotorControllerConfig rightMotorConfig = new SmartMotorControllerConfig(this)
      .withControlMode(ControlMode.OPEN_LOOP)
      .withGearing(gearing)
      .withIdleMode(MotorMode.COAST)
      .withMotorInverted(false)
      .withWheelDiameter(wheelDiameter)
      .withTelemetry("RightMotorMain", TelemetryVerbosity.LOW)
      .withFollowers(Pair.of(rightFollowerMotor, false));

  // SparkWrapper wraps each leader; followers are handled internally by the config.
  // DCMotor.getNEO(2) tells the simulator two NEOs are driving this side.
  private SmartMotorController leftMotorController  = new SparkWrapper(leftMotor, DCMotor.getNEO(2), leftMotorConfig);
  private SmartMotorController rightMotorController = new SparkWrapper(rightMotor, DCMotor.getNEO(2), rightMotorConfig);

  // DifferentialDrive applies its own squaring and deadband to the duty-cycle inputs,
  // so raw controller values are safe to pass through without pre-processing.
  private DifferentialDrive drive = new DifferentialDrive(leftMotorController::setDutyCycle,
                                                          rightMotorController::setDutyCycle);

  public DiffDriveSubsystem()
  {
    // stop() as default prevents the motors from holding their last output
    // if the driver command ends unexpectedly.
    setDefaultCommand(stop());
  }

  public Command stop()
  {
    return run(drive::stopMotor);
  }

  /**
   * Independent left/right speed control. Useful for autonomous routines that
   * need precise per-side voltage, or for testing individual sides.
   */
  public Command tankDrive(DoubleSupplier left, DoubleSupplier right)
  {
    return run(() -> drive.tankDrive(left.getAsDouble(), right.getAsDouble()));
  }

  /**
   * Single-stick arcade drive. xSpeed drives forward/backward; zRotation
   * turns in place. This is the default teleop binding.
   */
  public Command arcadeDrive(DoubleSupplier xSpeed, DoubleSupplier zRotation)
  {
    return run(() -> drive.arcadeDrive(xSpeed.getAsDouble(), zRotation.getAsDouble()));
  }

  @Override
  public void periodic()
  {
    leftMotorController.updateTelemetry();
    rightMotorController.updateTelemetry();
  }

  @Override
  public void simulationPeriodic()
  {
    leftMotorController.simIterate();
    rightMotorController.simIterate();
  }
}
